File: server/modules/auth/services/login-validation.ts
/**
 * Validação dos campos de login (e-mail/senha) — só forma/limites, sem whitelist de
 * domínio (isso é regra de cadastro, não de login).
 */
import {
  rustValidateLoginEmail,
  rustValidateLoginFieldsPresent,
  rustValidateLoginPassword
} from './auth-rust-bridge.js';

export type LoginValidationResult = { ok: true } | { ok: false; error: string };

export const EMAIL_ADDRESS_MAX_LENGTH = 50;
export const PASSWORD_MAX_LENGTH = 128;

export function validateLoginFieldsPresent(rawEmail: unknown, rawPassword: unknown): LoginValidationResult {
  const rust = rustValidateLoginFieldsPresent(rawEmail, rawPassword);
  if (rust) return rust;
  const emailStr = typeof rawEmail === 'string' ? rawEmail : '';
  const passwordStr = typeof rawPassword === 'string' ? rawPassword : '';
  const hasEmail = emailStr.trim().length > 0;
  const hasPassword = passwordStr.length > 0;
  if (!hasEmail && !hasPassword) {
    return { ok: false, error: 'Enter your email and password.' };
  }
  if (!hasEmail) {
    return { ok: false, error: 'Enter your email.' };
  }
  if (!hasPassword) {
    return { ok: false, error: 'Enter your password.' };
  }
  return { ok: true };
}

const FORBIDDEN_EMAIL_CHARS = /[<>'"\\]/;

export function validateLoginEmail(raw: unknown): LoginValidationResult {
  const rust = rustValidateLoginEmail(raw);
  if (rust) return rust;
  if (raw == null || typeof raw !== 'string') {
    return { ok: false, error: 'Enter your email.' };
  }
  const normalized = raw.trim().toLowerCase();
  if (!normalized) {
    return { ok: false, error: 'Enter your email.' };
  }
  if (normalized.length > EMAIL_ADDRESS_MAX_LENGTH) {
    return { ok: false, error: `Email may be at most ${EMAIL_ADDRESS_MAX_LENGTH} characters.` };
  }
  const at = normalized.lastIndexOf('@');
  if (at < 1 || at === normalized.length - 1) {
    return { ok: false, error: 'Invalid email.' };
  }
  const local = normalized.slice(0, at);
  const domain = normalized.slice(at + 1);
  if (
    !local ||
    !domain ||
    local.length + 1 + domain.length > EMAIL_ADDRESS_MAX_LENGTH ||
    domain.includes('..') ||
    domain.startsWith('.') ||
    domain.endsWith('.')
  ) {
    return { ok: false, error: 'Invalid email.' };
  }
  if (FORBIDDEN_EMAIL_CHARS.test(local) || FORBIDDEN_EMAIL_CHARS.test(domain)) {
    return { ok: false, error: 'Email contains disallowed characters.' };
  }
  return { ok: true };
}

export function validateLoginPassword(raw: unknown): LoginValidationResult {
  const rust = rustValidateLoginPassword(raw);
  if (rust) return rust;
  if (raw == null || typeof raw !== 'string') {
    return { ok: false, error: 'Enter your password.' };
  }
  if (raw.length > PASSWORD_MAX_LENGTH) {
    return { ok: false, error: `Password is too long (maximum ${PASSWORD_MAX_LENGTH} characters).` };
  }
  return { ok: true };
}

File: rust/genesis-core/src/auth/login_validation.rs
use super::constants::{EMAIL_ADDRESS_MAX_LENGTH, PASSWORD_MAX_LENGTH};
use super::types::PolicyResult;

fn forbidden_email_char(c: char) -> bool {
    matches!(c, '<' | '>' | '\'' | '"' | '\\')
}

pub fn validate_login_fields_present(
    raw_email: Option<&str>,
    raw_password: Option<&str>,
) -> PolicyResult {
    let email_str = raw_email.unwrap_or("");
    let password_str = raw_password.unwrap_or("");
    let has_email = !email_str.trim().is_empty();
    let has_password = !password_str.is_empty();
    if !has_email && !has_password {
        return PolicyResult::err("Enter your email and password.");
    }
    if !has_email {
        return PolicyResult::err("Enter your email.");
    }
    if !has_password {
        return PolicyResult::err("Enter your password.");
    }
    PolicyResult::ok()
}

pub fn validate_login_email(raw: Option<&str>) -> PolicyResult {
    let Some(raw) = raw else {
        return PolicyResult::err("Enter your email.");
    };
    let normalized = raw.trim().to_lowercase();
    if normalized.is_empty() {
        return PolicyResult::err("Enter your email.");
    }
    if normalized.len() > EMAIL_ADDRESS_MAX_LENGTH {
        return PolicyResult::err(format!(
            "Email may be at most {EMAIL_ADDRESS_MAX_LENGTH} characters."
        ));
    }
    let at = match normalized.rfind('@') {
        Some(i) if i >= 1 && i < normalized.len() - 1 => i,
        _ => return PolicyResult::err("Invalid email."),
    };
    let local = &normalized[..at];
    let domain = &normalized[at + 1..];
    if local.is_empty()
        || domain.is_empty()
        || local.len() + 1 + domain.len() > EMAIL_ADDRESS_MAX_LENGTH
        || domain.contains("..")
        || domain.starts_with('.')
        || domain.ends_with('.')
    {
        return PolicyResult::err("Invalid email.");
    }
    if local.chars().any(forbidden_email_char) || domain.chars().any(forbidden_email_char) {
        return PolicyResult::err("Email contains disallowed characters.");
    }
    PolicyResult::ok()
}

pub fn validate_login_password(raw: Option<&str>) -> PolicyResult {
    let Some(raw) = raw else {
        return PolicyResult::err("Enter your password.");
    };
    if raw.len() > PASSWORD_MAX_LENGTH {
        return PolicyResult::err(format!(
            "Password is too long (maximum {PASSWORD_MAX_LENGTH} characters)."
        ));
    }
    PolicyResult::ok()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fields_both_missing() {
        assert!(!validate_login_fields_present(Some(""), Some("")).is_ok());
    }

    #[test]
    fn email_ok() {
        assert!(validate_login_email(Some("a@b.com")).is_ok());
    }

    #[test]
    fn email_forbidden_chars() {
        assert!(!validate_login_email(Some("a<script>@b.com")).is_ok());
    }
}
